src/main/java/org/myrobotlab/math/MapperLinear.java

```diff
@@ -0,0 +1,237 @@
+package org.myrobotlab.math;
+
+import java.io.Serializable;
+
+import org.junit.runner.JUnitCore;
+import org.junit.runner.Result;
+import org.myrobotlab.logging.LoggerFactory;
+import org.myrobotlab.logging.LoggingFactory;
+import org.myrobotlab.service.WorkETest;
+import org.slf4j.Logger;
+
+public final class MapperLinear implements Serializable, MapperInterface {
+
+  private static final long serialVersionUID = 1L;
+  public final static Logger log = LoggerFactory.getLogger(MapperLinear.class);
+
+  // mapping range related
+  Double minX;
+  Double maxX;
+  Double minY;
+  Double maxY;
+
+  // clipping
+  Double minOutput;
+  Double maxOutput;
+
+  Boolean inverted = false;
+  Double maxInput;
+  Double minInput;
+
+  /**
+   * non-parameterized contructor for "un-set" mapper
+   * use the merge function to set "default" values from the motorcontroller
+   */
+  public MapperLinear() {
+  }
+  
+  public MapperLinear(double minX, double maxX, double minY, double maxY) {
+    map(minX, maxX, minY, maxY); 
+  }
+
+  /* (non-Javadoc)
+   * @see org.myrobotlab.math.MapperInterface#calcOutput(double)
+   */
+  @Override
+  final public double calcOutput(double in) {
+    if (minInput != null && in < minInput) {
+      log.warn("clipping input {} to {}", in, minInput);
+      in = minInput;
+    }
+    if (maxInput != null && in > maxInput) {
+      log.warn("clipping input {} to {}", in, maxInput);
+      in = maxInput;
+    }
+    
+    double c = minY + ((in - minX) * (maxY - minY)) / (maxX - minX);
+    
+    if (minOutput != null && c < minOutput) {
+      log.warn("clipping output {} to {}", c, minOutput);
+      return minOutput;
+    }
+    if (maxOutput != null && c > maxOutput) {
+      log.warn("clipping output {} to {}", c, maxOutput);
+      return maxOutput;
+    }
+    return c;
+  }
+
+  /* (non-Javadoc)
+   * @see org.myrobotlab.math.MapperInterface#calcOutputInt(int)
+   */
+  @Override
+  final public int calcOutputInt(int in) {
+    return (int) calcOutput(in);
+  }
+
+  /* (non-Javadoc)
+   * @see org.myrobotlab.math.MapperInterface#calcOutputInt(double)
+   */
+  @Override
+  final public int calcOutputInt(double in) {
+    return (int) calcOutput(in);
+  }
+
+  public double getMaxX() {
+    return maxX;
+  }
+
+  public double getMaxY() {
+    return maxY;
+  }
+
+  public double getMinX() {
+    return minX;
+  }
+
+  public double getMinY() {
+    return minY;
+  }
+
+  public boolean isInverted() {
+    return inverted;
+  }
+
+  public void setInverted(boolean invert) {
+    // change state ONLY if needed

```

Comment:
This should just set a boolean flag , and that flag should be honored in the calcOutput method.

Reply:
GAH !!!! .. copied over from original Mapper ... scraping clean ..

Reply:
actually .. I'll need to look at this .. one of the reasons this is messy, is because minY maxY were also used to describe minInput & maxInput ..  which has now been seperated..   minX maxX minY maxY are "now" ONLY describing  range mapping - NOT CLIPPING ...  Since they are separate we can choose what we want inverted.  One previous problem was min/max  was destroyed when a new range map was specified - which we certainly do not want to do.